### PR TITLE
Prompt refinement (discussion about if this should be merged or not)

### PR DESCRIPTION
**Pull request by haesleinhuepf**
This is a PR-draft because I'm not sure if we should merge this. **I'm curious about your opinions!**

This PR contains:
* [ ] a new test-case for the benchmark
  * [ ] I hereby confirm that NO LLM-based technology (such as github copilot) was used while writing this benchmark
* [ ] new dependencies in requirements.txt
  * [ ] The environment.yml file was updated using the command `conda env export > environment.yml`
* [ ] new generator-functions allowing to sample from other LLMs
* [ ] new samples (`sample_....jsonl` files)
* [ ] new benchmarking results (`..._results.jsonl` files)
* [ ] documentation update
* [ ] bug fixes 
* [x] test case / prompt modifications

Related github issue (if relevant):
* #111 
* #76 
* #79 

## Short description:
- In this PR I'm **modifying almost all prompts / docstrings** as discussed in [#111](https://github.com/haesleinhuepf/human-eval-bia/issues/111#issuecomment-2328390727_) and [#79](https://github.com/haesleinhuepf/human-eval-bia/issues/79#issuecomment-2220281539_) .
- I modified variable descriptions, e.g. from "image" to "image provided as numpy array". You can see all prompt modifications I did [here](https://github.com/haesleinhuepf/human-eval-bia/commit/2df0b8489b09c78dc95dfbb54950c83c36ae68c8?short_path=00dfcb0#diff-00dfcb0cb38cacfd1298499414374fdc3307c4280a6989a7c1d0a9c9033c8990) (these are many).
- I added sentences such as "The result is returned as numpy array" in case of function that returned an image, but the docstring didn't mention this.
- I also modified the reference code in four cases so that the specified type in the docstring is actually generated. E.g. I added `.to_list()` if the docstring mentioned that a result is returned as list. I also added `np.asarray(x)` to reference code in case the prompt mentioned that the result is a numpy array, but `x` wasn't of this type. To test if everything still works, I ran the evaluation for the reference code.

## How do you think will this influence the benchmark results?
- I tested how this influences the results by testing one commercial and one open-weight model. You can see the evaluation results [here](https://github.com/haesleinhuepf/human-eval-bia/blob/997c218cda360ff571ad46a009c4d22a64841917/demo/summarize_by_case.ipynb)
- For both LLMs the mean pass-rate improves by about 5%.
- This experiment costed about $7 OpenAI API usage (just in case anyone wants to try more modifications).

image

image

- This has subjectively positive impact on some test-cases (green frames) and negative impact on some other test cases (magenta frames). Notably, I see no impact on the test-cases the LLMs were all very bad in (bottom).
 
image

## Why do you think it makes sense to merge this PR?
- I'm not sure if we should merge this. There are pro's and cons.
- Pros:
  - By standardizing the prompts, we can potentially draw different conclusions about which LLM knows about which library for example.
  - The results are quantiatively better (but I'm not sure if this is a goal)
  - I presume the modifications are fair and do not give an advantage to specific models.
- Cons:
  - Merging PRs like this one turns our project from a "Benchmarking" project to "Prompt Engineering" project. This can lead us deep down into a rabbit hole. Many modifications are thinkable, see also #115 
  - Standardization also introduces a bias. This limits potential conclusions. We are only benchmarking if LLMs can solve bio-image analysis questions _if variable types are clearly defined_. I'm not sure if this was the original goal. This also somehow shrinks the target audience of the paper / LLM-usage to people who know what's a "numpy array". This actually goes a bit against our original goal: Measuring to what degree LLMs can help people with limited coding skills.
  - I suspected this modification might solve #76 but it obviously does not.

If we decide to merge this, I will remove all jsonl files in this branch and merge it into the development branch where we are also collecting new test-cases. The new cases there will also require the same docstring modification as done here. We will then later see in a full benchmark run how this impacts results.

I'm mildly against merging this. But if a majority says we should, I will do it.

## Poll

As I don't want to decide this alone, it would be great if everyone could use the 👍 and 👎  buttons below to express their opinion if we should merge this or not. If you think, we should only merge some of the modifications, please specify below which.

### CODE REVIEW

**Comment by ian-coccimiglio** [comment]
I don't think we should be pre-specifying desired object type within the prompts, so I'm a vote against merging. When I'm using an LLM, I find I ask more often for what I want, rather than how it ought to be.

Instead, I think we should be more concerned with what our benchmarks decide is a "pass" and "fail". This is a genuinely hard question, but I think it is what drives the majority of the counter-intuitive results from #76. My attempt at a solution heuristic is this: We should try to write acceptable solutions based on what would work for the modal question-asker. If our question asks to process a 2D image, and it fails because we were testing it on a 2D numpy array of type uint64, we should realize that that is rarely what a normal researcher would actually ever want to do.

I'll need to reflect a bit more, but also I have a few more fixes that should show how much mileage we can get out of better specified benchmarks rather than better specified prompts.

**Comment by tischi** [comment]
While I agree it would be cool to be able to support "leisurely written" prompts, I feel this would be a lot of work for us, because I guess that our testing code would need to become much more complicated.

I would vote for (i) merging the PR to as a straightforward way to improve our current framework, and (ii) have a separate issue (and branch) where motivated people could explore the very interesting question of how we could we could support more "leisurely written prompts".

Another thought: Maybe it is best practice to write prompts that are somewhat specific in terms of what you want as input and output types. And maybe, users that don't have this knowledge should not use LLMs to create code, because they will likely also not be able to judge the quality of the LLM's response?

**Comment by haesleinhuepf** [comment]
> I would vote for (i) merging the PR

Please use the thumbs-up button in the original post on the very top @tischi . It could otherwise be complicated to document the democracy in this thread 😉 Thanks

> and (ii) have a separate issue (and branch) where motivated people could explore the very interesting question of how we could we could support more "leisurely written prompts".

I like this idea a lot actually.

**Comment by tischi** [comment]
> Please use the thumbs-up button

I did and I noticed that you voted _both_ thumbs up and down, on purpose?

**Comment by haesleinhuepf** [comment]
Yes, because I wanted to have both buttons visible there.

**Comment by haesleinhuepf** [comment]
Hi  @jkh1  @nscherf @royerloic @pr4deepr @marabuuu @shinzlet @ilan-theodoro @TeunHuijben @JoOkuma,

what's your opinion on this modification?

Big thanks for your time!

Best,
Robert

**Comment by jkh1** [comment]
On the one hand, I feel it's good to have standardized prompts that allow better evaluation of the models as well as trying to increase pass rates. On the other hand, the original idea was to evaluate output to requests by "naive" users. I think we've learned a lot from this exercise and we should devote a section of the paper to it or even to all the improvements that have been made since the first version. I believe it would be very valuable for readers (and may raise the profile of the paper) to see the types of errors that have to be dealt with at different levels, from how the models fail in different cases to how to properly parse their outputs. 
So in short, I support merging if we extensively discuss it in the paper.

**Comment by pr4deepr** [comment]
I like these arguments above and I believe that _Introducing a level of standardization makes it a better domain-specific benchmarking exercise._

But, my concerns are same as what Robert mentioned. 
**Are we biasing our benchmarking towards making the LLMs perform better? Is this a benchmarking project anymore?**

**Where do we draw the line?**

For example, if you look at [summarize error messages](https://github.com/haesleinhuepf/human-eval-bia/blob/main/demo/summarize_error_messages.ipynb), top errors messages are related to OpenCV. 
What if we changed our prompt to get the LLM to use scikit image instead? It is possible that training data for LLMs have many  OpenCV examples for general image processing, and scikit image examples for bioimage analysis workflows.
I would argue that we need to be more specific and say, use scikit image library. 
What if this prompt increased the pass rate? Its an example of how to make a better prompt.
I'm testing this at the moment.



I am pro merging, BUT,

- discuss importance of this and why users should be specific.  Show objective data  (old prompts vs new). This is quite an important result.
- potentially have the refined prompt as a second graph in the paper in addition to the original one -> Simple prompts vs domain specific prompts

For long term maintenance and to reduce workload, we can have the refined prompts in the main repo, and point users to a different branch for the older prompts.

Cheers
Pradeep

**Comment by pr4deepr** [comment]
> I think we should be more concerned with what our benchmarks decide is a "pass" and "fail".

Agree with this too

**Comment by pr4deepr** [comment]
minor improvement when I just added the term "Use scikit image library." You can find [what I tested here](https://github.com/pr4deepr/human-eval-bia/tree/prompt_refinement/test_cases). Keep in mind this was a quick test.

image

image


examples where I added the above prompt..

image

This could look different if I had said, use scikit image library where possible or something similar.

**Comment by haesleinhuepf** [comment]
> What if we changed our prompt to get the LLM to use scikit image instead?

Since this project I'm also wondering if I _should_ use OpenCV more?

**Comment by jkh1** [comment]
> I'm also wondering if I should use OpenCV more  

I've also been wondering why bioimage analysts have been ignoring decades of work and tooling from the computer vision field :confused:

**Comment by jkh1** [comment]
>  Is this a benchmarking project anymore?
This turned into quite some exploratory work and benchmarking is now only just part of it so I guess it depends on where we want to put the emphasis.

**Comment by ian-coccimiglio** [comment]
Re: OpenCV and Bioimage analysis, for me (in my bioimage analysis journey) the decisions often came down to errors and error-handling. OpenCV is (I think) much more strict about typing than most typical python imaging libraries are. In this project here, about 1200/8800 errors are related to an OpenCV exception having been thrown.

>This turned into quite some exploratory work and benchmarking is now only just part of it so I guess it depends on where we want to put the emphasis.

I can see three really compelling use-cases for this project for others.
- Novices may want to use this project to get a sense of "which bioimaging tasks can LLMs do easily"
- Researchers may want to use this project to get a sense of "which bioimaging tasks can LLMs not do easily"
- Historians may want to use this project as a time capsule indicating the level of rudimentary sophistication of our LLMs in the year 2024 :smile: 

So I quite like the benchmarking & I hope that remains a focus.
